File: 637-average-of-levels-in-binary-tree/637-average-of-levels-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> ans=new ArrayList<>();
        Queue<TreeNode> q=new LinkedList<>();
        q.add(root);
        while(q.size()>0){
            int sz=q.size();
            int siz=sz;
            Double avg=0.00;
            while(siz-->0){
                TreeNode curr=q.remove();
                if(curr.left!=null)q.add(curr.left);
                if(curr.right!=null)q.add(curr.right);
                avg+=curr.val;
            }
            avg/=sz;
            ans.add(avg);
        }
        return ans;
    }
}

File: 637-average-of-levels-in-binary-tree/637-average-of-levels-in-binary-tree.cpp
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        vector<double> v;
        while(!q.empty()){
            int n = q.size();
            double d=0;                        
            for(int i=0;i<n;i++){
                TreeNode* node = q.front();
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
                q.pop();
                d += node->val;
            }
            v.push_back(d/n);
        }
        return v;
    }
};


// Same approach:
// class Solution {
// public:
//     vector<double> ans;
//     void getAverage(TreeNode *root){
//         queue<TreeNode*> q;
//         q.push(root);
//         while(!q.empty()){
//             double size = q.size();
//             double sum = 0;
//             for(int i = 0; i < size; i++){
//                 auto node = q.front();
//                 q.pop();
//                 sum += node->val;
//                 if(node->left)
//                     q.push(node->left);
//                 if(node->right)
//                     q.push(node->right);
//             }
//             ans.push_back(sum / size);
//         }
//     }
    
//     vector<double> averageOfLevels(TreeNode* root) {
//         getAverage(root);
//         return ans;
//     }
// };
